keps/0020-controller-overhaul.md

```diff
@@ -0,0 +1,160 @@
+
+### Expose custom HTTP API from controller-manager to allow receiving commands
+
+Manager binary with controllers (that is run within the pod in the cluster) would in this case have to also expose HTTP API and we’ll have to expose a kubernetes service with this custom API. All clients (CLI/user scripts) will talk to this API every time they want to start a new plan. The API could be as simple as 
+
+HTTP POST `/start?planName=deploy`
+
+The validation whether a plan can be started or not would have to be part of the HTTP endpoint implementation.
+
+#### Pros
+
+The complexity of the implementation is still pretty managable and lower than having a full-featured sub-resource
+
+#### Cons
+
+Have to maintain a separate API (with API documentation)
+It is a non-kubernetes API, so no clients for various languages and no kubectl support (although if we keep the API that simple, almost a CURL can do the trick)
+More complex manager pod deployment with additional service
+How do we manage RBAC?
+
+### Have a separate CRD just for maintaining commands to execute a plan
+
+This would basically mean going half-way back to state pre KEP-18 where we had a separate CRD that was created for every plan that was executed. In this implementation we could have a much leaner CRD capturing only the requests to execute a plan with very lean status that does not duplicate the instance's status. Could be as simple as:
+
+```
+apiVersion: kudo.dev/v1beta1
+kind: PlanExecutionRequest
+Spec:
+  planName: deploy
+Status:
+  Started: true
+```
+
+The question is whether we would add a Status into that CRD. If not, it would be hard to tell the client if the execution was even possible and how it ended (in comparison to the API solution where we can validate as part of HTTP response and even return some unique ID for tracking). If we decide to include status, we’re back in pre-KEP-18 world where keeping those two Statuses in sync is hard in kubernetes world with no transactions and both CRDs being reconciled in their own loops.
+
+But at the same time this seems to be the way people are leaning toward when coming to similar use-cases. The discussion under [this issue](https://github.com/kubernetes/kubernetes/issues/72637#issuecomment-515566586) is kind of touching on the topic of Request objects. What this [KEP](https://github.com/metal3-io/metal3-docs/pull/48/files) proposes for reboot is also kind of similar to what we’re trying to do. Alongside what they’re proposing we might not even need a controller for this CRD and we could treat it as “queue” meaning every request will be fulfilled so e.g. when you create a request and a plan is running, instance will pick up the next request when it has a capacity to run another plan.
+
+If we do that, it would mean:
```

Comment:
I need to sleep on it but this seems like the least evil of the options, considering all the pros and cons 🙈

Reply:
Yeah ... I recommend reading that issue and KEP I linked, that basically kind of changed my mind... but even in that KEP they make it obvious that you should not have another controller for that other CRD so we won't repeat all the problems of PlanExecution hopefully...

Reply:
I went through both yesterday and I still think that a custom CR subresource would be a better fit. However, as you said, the important thing is that we don't end up with two controllers here.

Reply:
just to be sure - when you're saying subresource, you mean `PlanExecutionRequest` CRD, right?

I'll wait for @gerred or someone else to chime in and if we feel like this is the way to go I'll try to outline the implementation a bit more and then we can work on that...

Reply:
I'm leaning to PlanExecutionRequest as well.

Reply:
> just to be sure - when you're saying subresource, you mean `PlanExecutionRequest` CRD, right

no, I meant the `Instance` resource. However, as Gerred mentioned below, this will blow up the etcd if we try to keep a history (for audit and what not).